### ISSUE TITLE
Allow once-per-parameter initialization

### ISSUE THREAD

**Issue by anderspapitto** [bheisler/criterion.rs#247]
I'm using `bench_function_over_inputs`, and I need to do one-time initialization, per input. Effectively the inputs I'm ranging are are configurations for a large data structure, which I then benchmark quick operations against.

Without the 'per input' aspect, I've found recommendations to simply initialize the data at the top level, and them move it into the outer closure. That won't work in this case because the input value isn't available outside of the outer closure.

Some example pseudocode
```
fn my_bench(c: &mut Criterion) {
    c.bench_function_over_inputs(move |b, i| {

        // this depends on `i` but I only want it called once
        let input = generate_input(i);

        b.iter(|| my_function(input))
    }, &[1, 2, 4, 8]);
}
```

Here are the earlier discussions I've looked at.

https://www.reddit.com/r/rust/comments/8uj5oa/criterion_benchmark_with_onetime_setup_also_how/e1hd7he/
https://github.com/bheisler/criterion.rs/issues/225#issuecomment-429365770

**Comment by bheisler** [comment]
Hey, thanks for trying Criterion.rs, and thanks for the suggestion!

Yeah, this has come up before. Would it be possible to generate all of the inputs up-front and pass those as the parameters to the benchmark? It would be kind of tricky to add something like this into the Criterion.rs API as it stands; supporting this well would probably require a breaking change. I'll have to think about this some more when I get started on 0.3.0, but I can't promise anything at this point.

**Comment by anderspapitto** [comment]
@bheisler it would be possible with enough contortion, but at that point it's going to feel like a much less natural use of bench_function_over_inputs, since I'll have to duplicate the parameter list elsewhere and have an extra loop of setup.

**Comment by bheisler** [comment]
I think this would be solved by the changes described in #261. Does that seem right to you?

**Comment by bheisler** [comment]
I think this should be do-able now using the `BenchmarkGroup` struct in 0.3.0. I'm going to close this for now; please re-open if that doesn't solve your problem.
